slaclab/lcls-tools
# Reorganise LCLS-Tools modules for unit tests

## Summary of Changes

The below reorganization may have an effect on codebases that utilize these modules. Please review your code after reading through these changes.

### Source Code Reorganization

As part of #84 it was decided that some cleaning up of the module structure was needed along with providing the unit-test directory structure.

In the top level directory, we still have:

```
lcls-tools
  -> docs
  -> examples
  -> lcls_tools
  ----> common
  ----> normalconducting
  ----> superconducting  
```

However, some submodules of the `common` module have now been reorganized:
```
-----> common
--------> beam_calcs (still the same)
--------> beam (not implemented yet)
-----------> transverse (not implemented yet)
-----------> longitudinal (not implemented yet) 
--------> controls
-----------> pyepics (formerly pyepics_tools)
--------> data_analysis (still contains archiver.py)
-----------> fitting (formerly math_tools)
--------> devices (still the same)
--------> frontend (formerly pydm_tools)
-----------> display (contains displayUtils, now known as util)
-----------> plotting (contains pydmPlotUtil, now known as util)
-----------> widgets (contains magnet_template.ui etc. to be use for more complex widget sharing)
--------> image_processing (still the same, needs to be dissected a bit more)
--------> logger (still the same)
--------> matlab2py (used for common dataset conversions from matlab applications to python types)
-----------> cor_plot (for correlation plot matlab files)
-----------> emit_scan (for emittance scan matlab files)
-----------> image (for images stored stored as matlab files)
-----> normalconducting (still the same)
-----> superconducting (still the same)
```

### Unit Test Directory Structure

We are using the `unittest` module in python standard library to conduct the testing of the lcls-tools module. To allow `unittest` to find the tests automatically, we must have a specified directory structure that mirrors the top-level source code directories.

Some of the previous unit tests were mixed in with the source code and they have been separated out into the test area and the associated datasets have been added to the `tests/datasets` sub-folders

The aim of this PR is not to make the tests pass, but to give us the ability to run all of the tests via `unittest`.

### To Test

- Go to the top level of this repository
- Run `python3 -m unittest discover'
- You should see that: 
  - 38 tests ran!
  - The majority of tests are skipped due to not finding the dataset files
  - 2 tests fail due to the expected values being tested not matching the dataset files
